NKL/Encode.lean

```diff
@@ -0,0 +1,331 @@
+
+-/
+
+namespace NKL
+
+-- All of the encode function are pure; decoding uses an instance of EStateM.
+
+abbrev DecodeM := EStateM String ByteArray.Iterator
+
+def decode' (f : DecodeM a) (ba : ByteArray) : Option a :=
+  EStateM.run' f ba.iter
+
+def decode (f : DecodeM a) (ba : ByteArray) : Except String a :=
+  match EStateM.run f ba.iter with
+  | .ok x _ => .ok x
+  | .error s _ => .error s
+
+def decodeFile (f : DecodeM a) (path : System.FilePath) : IO a := do
+ let buf <- IO.FS.readBinFile path
+ match decode f buf with
+ | .ok x => return x
+ | .error s => throw $ IO.userError s
+
+private def next : DecodeM UInt8 := do
+  let it <- get
+  if it.atEnd then throw "unexpected EOF"
+  let b := it.curr
+  set it.next
+  return b
+
+-- TODO: this could be more efficient if we complicate the monad state
+private def take (n : Nat) : DecodeM ByteArray :=
+  let rec f (ba : ByteArray) (n : Nat) : DecodeM ByteArray :=
+    match n with
+    | .zero => return ba
+    | .succ n => next >>= fun b => f (ba.push b) n
+  f (.mk #[]) n
+
+------------------------------------------------------------------------------
+-- Nats are encoded as u64, little-endian
+-- TODO: if the nat is too large, it will be silently truncated.
+
+private def encNat (n : Nat) : ByteArray := .mk
+  #[ UInt8.ofNat $ n >>> 0
+   , UInt8.ofNat $ n >>> 8
+   , UInt8.ofNat $ n >>> 16
+   , UInt8.ofNat $ n >>> 24
+   , UInt8.ofNat $ n >>> 32
+   , UInt8.ofNat $ n >>> 40
+   , UInt8.ofNat $ n >>> 48
+   , UInt8.ofNat $ n >>> 56
+   ]
+
+private def decNat : DecodeM Nat := do
+  let nat8 : DecodeM Nat := next >>= fun x => return x.toNat
+  return (<- nat8) <<< 0  |||
+         (<- nat8) <<< 8  |||
+         (<- nat8) <<< 16 |||
+         (<- nat8) <<< 24 |||
+         (<- nat8) <<< 32 |||
+         (<- nat8) <<< 40 |||
+         (<- nat8) <<< 48 |||
+         (<- nat8) <<< 56
+
+#guard decode' decNat (.mk #[1,2,3,4,5,6,7]) == none
+#guard decode' decNat (.mk #[0,1,0,0,0,0,0,0]) == some 256
+#guard decode' decNat (encNat 0) == some 0
+#guard decode' decNat (encNat 1) == some 1
+#guard decode' decNat (encNat 256) == some 256
+
+------------------------------------------------------------------------------
+-- Ints are encoded as a sign tag followed by a Nat
+-- TODO: Int64 should be in standard library soon, revisit when available
+
+private def tag (t : UInt8) : List ByteArray -> ByteArray :=
+  List.foldl ByteArray.append (.mk #[t])
+
+private def encInt (i : Int) : ByteArray :=
+  match i with
+  | .ofNat n => tag 0 [encNat n]
+  | .negSucc n => tag 1 [encNat n]
+
+private def decInt : DecodeM Int := do
+  let sign <- next
+  let n <- decNat
+  bif sign == 0 then
+    return (.ofNat n)
+  else
+    return (.negSucc n)
+
+private def decInt' : DecodeM (Bool × Nat) := do
+  let sign <- next
+  let n <- decNat
+  return (sign == 1, n + sign.toNat)
+
+local instance : BEq ByteArray where
+  beq a b := a.data == b.data
+
+#guard encInt 1    == .mk #[0, 1, 0, 0, 0, 0, 0, 0, 0]
```

Comment:
Pedantic comment: imo it's a bit unusual to have the sign bit as the 0th bit on a little-endian representation. The sign seems like the "most important" bit.

Reply:
Indeed, also 9 bytes is a too many bytes :)
Having Int64 in lean will fix this.